src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs

```diff
@@ -25,21 +25,51 @@ public static class InlineExpressionMappingBuilder
         }
 
         var methodSyntax = methodSyntaxRef.GetSyntax();
-        if (methodSyntax is not MethodDeclarationSyntax { ExpressionBody: { } body, ParameterList.Parameters: [var sourceParameter] })
+
+        var bodyExpression = methodSyntax switch
+        {
+            // => expression
+            MethodDeclarationSyntax { ExpressionBody: { } body, ParameterList.Parameters: [var sourceParameter1] } => body.Expression,
+
+            // { return expression; }
+            MethodDeclarationSyntax { Body.Statements: [ReturnStatementSyntax singleStatement] } => singleStatement.Expression,
+
+            // { var dest = expression; return dest; }
+            MethodDeclarationSyntax
+            {
+                Body.Statements: [
+                    LocalDeclarationStatementSyntax
+                    {
+                        Declaration.Variables: [VariableDeclaratorSyntax { Initializer: { } variableInitializer } variableDeclarator]
+                    },
+                    ReturnStatementSyntax { Expression: IdentifierNameSyntax identifierName }
+                ]
+            } when identifierName.Identifier.Value == variableDeclarator.Identifier.Value
+                => variableInitializer.Value,
+
+            _ => null
+        };
+        if (bodyExpression == null)
         {
             ctx.ReportDiagnostic(DiagnosticDescriptors.QueryableProjectionMappingCannotInline, mapping.Method);
             return null;
         }
 
         var semanticModel = ctx.Compilation.GetSemanticModel(methodSyntax.SyntaxTree);
         var inlineRewriter = new InlineExpressionRewriter(semanticModel, ctx.FindNewInstanceMapping);
-        var bodyExpression = (ExpressionSyntax?)body.Expression.Accept(inlineRewriter);
+        bodyExpression = (ExpressionSyntax?)bodyExpression.Accept(inlineRewriter);
         if (bodyExpression == null || !inlineRewriter.CanBeInlined)
         {
             ctx.ReportDiagnostic(DiagnosticDescriptors.QueryableProjectionMappingCannotInline, mapping.Method);
             return null;
         }
 
+        if (methodSyntax is not MethodDeclarationSyntax { ParameterList.Parameters: [var sourceParameter] })
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.QueryableProjectionMappingCannotInline, mapping.Method);
+            return null;
+        }
```

Comment:
This is not covered by tests and I can't think of a use-case, is it really required?

Reply:
`UserImplementedInlinedExpressionMapping.ReplaceSource` needs `sourceParameter` later. I moved this up a bit in order to save redundant `MethodDeclarationSyntax` checks.